runtime/src/accounts_db.rs

```diff
@@ -249,6 +249,17 @@ impl AccountStorageEntry {
         self.accounts.flush()
     }
 
+    pub fn all_existing_accounts(&self) -> Vec<StoredAccount> {
+        self.accounts.accounts(0)
+    }
+
+    pub fn restore_account_count(&self) {
+        let mut count_and_status = self.count_and_status.write().unwrap();
+        let new_count = self.all_existing_accounts().len();
```

Comment:
I don't think this is right actually. The account could be masked by an account on another fork. This should only count accounts where the latest version is represented in this append vec.

Reply:
Thanks for early preview!

Yes, your concern is right. 

To make sure not to use older versions of accounts, we gradually decrease the count [here](https://github.com/solana-labs/solana/pull/7013/files#diff-2099c5256db4eb5975c8834af38f6456R1186) by calling `remove_account`, once increased to the full count of stored accounts here. I think calculating the right count here is difficult because this method can't tell there are any other newer representation of accounts if any.

As for the freshness check of the actual account contents like the counts, we realize it by replacing older account with new one  in chronological order when generating the index as done by [the original implementation](https://github.com/solana-labs/solana/pull/7013/files#diff-2099c5256db4eb5975c8834af38f6456L1120). The handling is similar to counts.

Yeah, code is a bit difficult to read. Maybe I can add comment to this method, or moreover directly embed the code into `generate_index` as this is the sole caller of this (and should be in the future). Even if we completely transition not to serialize the count, I'll leave a comment explaining that.